README.md

# BBTools
---

This module makes [BBTools](https://jgi.doe.gov/data-and-tools/bbtools/) available in KBase.  

It currently (5/2/2018) only exposes RQCFilter for use as either a standalone tool or as part of other pipelines, but other tools (BBMap, BBDuk, etc.) will be available over time.

## Usage
RQCFilter can be run either as an App from the Narrative Interface, or as a local function from other KBase Apps. Along with the configured parameters, it always uses the following set of arguments:
```
-Xmx24g barcodefilter=f sketch=f mapk=13
```
And a provided set of JGI Reference Data, kindly provided by Brian Bushnell.

To run it locally, first install this module with:
```
kb-sdk install BBTools
```
Then you can run `run_RQCFilter_local` with the following steps (Python below, Java and Perl will differ slightly).

Note that not all RQCFilter parameters are available right now. See [BBTools.spec](https://github.com/briehl/BBTools/blob/master/BBTools.spec#L12-L68) for a list.

```
# instantiate the client
from BBTools.BBToolsClient import BBTools
bbtools = BBTools(callback_url)   # see the SDK docs about finding the callback_url.

# set up input files
rqc_filter_input = {
    "reads_file": "/path/to/interleaved/fastq"
}
# or, if you want to use a KBase Workspace UPA for your reads object:
rqc_filter_input = {
    "reads_library_ref": "xx/yy/zz"
}

# set up parameters (example below, there are many more options, see BBTools.spec)
rqc_filter_params = {
    "qtrim": "rl",
    "maxns": 3,
    "minlength": 40
}

# run the local RQCFilter function
result = bbtools.run_RQCFilter_local(rqc_filter_input, rqc_filter_params)
```

The result of an RQCFilter run is a structure with the following keys:
* output_directory - the absolute path to the directory containing all output from the RQCFilter run.
* run_log - the absolute path to the generated log from RQCFilter (what's output to stderr)
* filtered_fastq_file - the absolute path to the generated fastq file that has been filtered according to the user's criteria.

All of the output directory info should be available from your SDK module's file system.

# How to Update
There's a few caveats for updating the version of BBTools being used here. Here's the steps to follow.
1. Update Dockerfile.  
The version downloaded is embedded in the `Dockerfile` for this repo, line 29. Bump that version to whatever you need.
2. Update App Spec.  
The version is also added to `ui/narrative/methods/RQCFilter/display.yaml`, line 4. Update that there to display the version of BBTools being run.
3. Update kbase.yml  
The `module-version` in `kbase.yml` should also be bumped. It follows semantic versioning, so if you're just bumping the BBTools version, then a patch upgrade is most appropriate.
4. (optional) Update reference data.  
If the reference data changes - either in the large tar file hosted at NERSC, or what's packaged with bbmap, this gets slightly more complex. The file `scripts/load_reference_data.sh` works outside of the module compile process, so it maintains its own version of things. There's a version of BBTools embedded there as well, so that will need to change if the bundled data is updated. Finally, you'll have to increase the value of the `data-version` line in `kbase.yml` for the reference data to be re-downloaded.


RELEASE_NOTES.md
### Version 1.1.0
__Changes__
- updated BBMap to 38.94 (released 2021-09-30)

### Version 1.0.1
__Changes__
- Added config options and html results to BBMap
- Update the documentation in the display.yaml file

### Version 1.0.0
__Changes__
- Added BBMap v38.73 (released 2019-11-14)
- updated RQCFilter to v38.73

### Version 0.5.0
__Changes__
- RQCFilter v38.22 supported


lib/BBTools/BBToolsClient.py
# -*- coding: utf-8 -*-
############################################################
#
# Autogenerated by the KBase type compiler -
# any changes made here will be overwritten
#
############################################################

from __future__ import print_function
# the following is a hack to get the baseclient to import whether we're in a
# package or not. This makes pep8 unhappy hence the annotations.
try:
    # baseclient and this client are in a package
    from .baseclient import BaseClient as _BaseClient  # @UnusedImport
except ImportError:
    # no they aren't
    from baseclient import BaseClient as _BaseClient  # @Reimport


class BBTools(object):

    def __init__(
            self, url=None, timeout=30 * 60, user_id=None,
            password=None, token=None, ignore_authrc=False,
            trust_all_ssl_certificates=False,
            auth_svc='https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login'):
        if url is None:
            raise ValueError('A url is required')
        self._service_ver = None
        self._client = _BaseClient(
            url, timeout=timeout, user_id=user_id, password=password,
            token=token, ignore_authrc=ignore_authrc,
            trust_all_ssl_certificates=trust_all_ssl_certificates,
            auth_svc=auth_svc)

    def run_BBMap(self, io_params, run_params, context=None):
        """
        :param io_params: instance of type "BBMapAppParams" (BBMap App IO) ->
           structure: parameter "workspace_name" of type "workspace_name" (A
           workspace_name - alphanumeric + '.' + '_' + '-' only permitted
           characters), parameter "in_assembly_refs" of list of type
           "data_obj_ref" (A data_obj_ref - address of form
           'WS_NAME/OBJ_NAME', 'WS_NAME/OBJ_NAME/VERSION', or 'WS_ID/OBJ_ID',
           or 'WS_ID/OBJ_ID/VERSION'), parameter "in_readslib_ref" of type
           "data_obj_ref" (A data_obj_ref - address of form
           'WS_NAME/OBJ_NAME', 'WS_NAME/OBJ_NAME/VERSION', or 'WS_ID/OBJ_ID',
           or 'WS_ID/OBJ_ID/VERSION'), parameter "out_obj_name" of type
           "data_obj_name" (A data_obj_name - alphanumeric + '.' + '_' + '-'
           only permitted characters)
        :param run_params: instance of type "BBMapParams" (BBMap method (App
           and Local)) -> structure: parameter "get_mapped_reads" of String,
           parameter "get_unmapped_reads" of String, parameter "get_bam" of
           String, parameter "input_parameter_suite" of String, parameter
           "use_modulo" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1)), parameter "speed_mode" of String, parameter
           "min_id" of Double, parameter "bandwidth" of String, parameter
           "min_hits" of Long, parameter "kmer_len" of Long, parameter
           "max_indel" of Long, parameter "strict_max_indel" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "subfilter_thresh" of Long, parameter "delfilter_thresh"
           of Long, parameter "require_correct_strand" of type "boolean" (A
           boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "perfect_mode" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1)), parameter "semiperfect_mode" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "qual_score_mode" of Long
        :returns: instance of type "BBMapAppOutput" -> structure: parameter
           "report_name" of String, parameter "report_ref" of String,
           parameter "run_command" of String, parameter "coverage" of list of
           mapping from String to mapping from String to Double, parameter
           "read_align_stats" of list of mapping from String to Double
        """
        return self._client.call_method('BBTools.run_BBMap',
                                        [io_params, run_params], self._service_ver, context)

    def run_BBMap_local(self, io_params, run_params, context=None):
        """
        :param io_params: instance of type "BBMapLocalParams" (BBMap Local
           IO) -> structure: parameter "in_assembly_paths" of list of type
           "file_path" (A file_path - absolute path to a file), parameter
           "in_readslib_path" of type "file_path" (A file_path - absolute
           path to a file), parameter "out_file_basename" of String
        :param run_params: instance of type "BBMapParams" (BBMap method (App
           and Local)) -> structure: parameter "get_mapped_reads" of String,
           parameter "get_unmapped_reads" of String, parameter "get_bam" of
           String, parameter "input_parameter_suite" of String, parameter
           "use_modulo" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1)), parameter "speed_mode" of String, parameter
           "min_id" of Double, parameter "bandwidth" of String, parameter
           "min_hits" of Long, parameter "kmer_len" of Long, parameter
           "max_indel" of Long, parameter "strict_max_indel" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "subfilter_thresh" of Long, parameter "delfilter_thresh"
           of Long, parameter "require_correct_strand" of type "boolean" (A
           boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "perfect_mode" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1)), parameter "semiperfect_mode" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "qual_score_mode" of Long
        :returns: instance of type "BBMapLocalOutput" -> structure: parameter
           "mapped_reads_paths" of list of type "file_path" (A file_path -
           absolute path to a file), parameter "unmapped_reads_paths" of list
           of type "file_path" (A file_path - absolute path to a file),
           parameter "bam_paths" of list of type "file_path" (A file_path -
           absolute path to a file), parameter "coverage_paths" of list of
           type "file_path" (A file_path - absolute path to a file),
           parameter "output_directory" of type "file_path" (A file_path -
           absolute path to a file), parameter "run_log" of type "file_path"
           (A file_path - absolute path to a file), parameter "run_command"
           of String
        """
        return self._client.call_method('BBTools.run_BBMap_local',
                                        [io_params, run_params], self._service_ver, context)

    def run_RQCFilter_app(self, io_params, run_params, context=None):
        """
        :param io_params: instance of type "RQCFilterAppParams" (Parameters
           for the Narrative App version of RQCFilter. read_library_ref - UPA
           for the read library to filter. output_workspace_name - name of
           the workspace to put the output reads library and report.
           output_library_name - name of the Reads library object produced by
           the app.) -> structure: parameter "read_library_ref" of String,
           parameter "output_workspace_name" of String, parameter
           "output_library_name" of String
        :param run_params: instance of type "RQCFilterParams" (Contains all
           parameters for the RQCFilter program, EXCEPT for the inputs and
           outputs. Those are added specifically by each function. This lets
           us describe them separately for the local function that works
           mainly against the file system and the app that mainly works
           against the Workspace. This doesn't cover all of the 110+
           parameters provided by rqcfilter. Those not listed here are left
           as default values, except sketch=f (as that sends data to JGI
           servers for processing), barcodefilter=f, and mapk=13. Notes below
           are taken from the help output from rqcfilter2.sh ver 38.00
           Parameters (format = param name - default - description):
           --------------------------------------------------------- library
           - frag - should be one of 'frag', 'clip', 'lfpe', or 'clrs'.
           Adapter trimming parameters: ----------------------------
           trimfragadapter - f - Trim all known Illumina adapter sequences,
           including TruSeq and Nextera. Quality trimming parameters:
           ---------------------------- qtrim - f - Trim read ends to remove
           bases with quality below minq. Performed AFTER looking for kmers.
           Values: rl (trim both ends), f (neither end), r (right end only),
           l (left end only). trimq - 10 - Trim quality threshold.  Must also
           set qtrim for direction, will be ignored if qtrim=f maxns - 0 -
           Reads with more Ns than this will be discarded. minavgquality - 5
           - (maq) Reads with average quality (before trimming) below this
           will be discarded. minlength - 45 - (ml) Reads shorter than this
           after trimming will be discarded.  Pairs will be discarded only if
           both are shorter. mlf - 0.333 - (minlengthfraction) Reads shorter
           than this fraction of original length after trimming will be
           discarded. Mapping parameters (for vertebrate contaminants):
           ------------------------------------------------- removemouse - f
           - (mouse) Remove mouse reads via mapping. removecat - f - (cat)
           Remove cat reads via mapping. removedog - f - (dog) Remove dog
           reads via mapping. removehuman - f - (human) Remove human reads
           via mapping. Microbial contaminant removal parameters:
           ----------------------------------------- removemicrobes - f -
           (microbes) Remove common contaminant microbial reads via mapping,
           and place them in a separate file. taxlist - emptylist - (tax)
           Remove these taxa from the database before filtering.  Typically,
           this would be the organism name or NCBI ID, or a comma-delimited
           list.  Organism names should have underscores instead of spaces,
           such as Escherichia_coli. Filtering parameters (for artificial and
           microbial contaminants):
           -----------------------------------------------------------------
           rna - f - Remove reads containing RNA-specific artifacts. phix - t
           - Remove reads containing phiX kmers. Clumpify parameters:
           -------------------- clumpify - f - Run clumpify. dedupe - f -
           Remove duplicate reads. opticaldupes - f - Remove optical
           duplicates (Clumpify optical flag). Other processing parameters:
           ---------------------------- khist - f - Set to true to generate a
           kmer-frequency histogram of the output data. (included in report
           in the app, as a file in local function) Memory requirements
           (DON'T EXPOSE THESE TO APPS):
           ------------------------------------------------- maxmem - 50 -
           Set maximum memory flag for RQCFilter to try to allocate. Should
           be an integer, in GB.) -> structure: parameter "library" of
           String, parameter "trimfragadapter" of type "boolean" (A boolean -
           0 for false, 1 for true. @range (0, 1)), parameter "qtrim" of
           String, parameter "trimq" of Long, parameter "maxns" of Long,
           parameter "minavgquality" of Long, parameter "minlength" of Long,
           parameter "mlf" of Double, parameter "removemouse" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "removecat" of type "boolean" (A boolean - 0 for false,
           1 for true. @range (0, 1)), parameter "removedog" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "removehuman" of type "boolean" (A boolean - 0 for
           false, 1 for true. @range (0, 1)), parameter "removemicrobes" of
           type "boolean" (A boolean - 0 for false, 1 for true. @range (0,
           1)), parameter "taxlist" of list of String, parameter "rna" of
           type "boolean" (A boolean - 0 for false, 1 for true. @range (0,
           1)), parameter "phix" of type "boolean" (A boolean - 0 for false,
           1 for true. @range (0, 1)), parameter "clumpify" of type "boolean"
           (A boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "dedupe" of type "boolean" (A boolean - 0 for false, 1 for true.
           @range (0, 1)), parameter "opticaldupes" of type "boolean" (A
           boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "khist" of type "boolean" (A boolean - 0 for false, 1 for true.
           @range (0, 1)), parameter "maxmem" of Long
        :returns: instance of type "RQCFilterAppOutput" -> structure:
           parameter "report_name" of String, parameter "report_ref" of
           String, parameter "run_command" of String
        """
        return self._client.call_method('BBTools.run_RQCFilter_app',
                                        [io_params, run_params], self._service_ver, context)

    def run_RQCFilter_local(self, io_params, run_params, context=None):
        """
        :param io_params: instance of type "RQCFilterLocalParams" (Parameters
           for local version of RQCFilter. read_library_ref - UPA for the
           read library to filter. -OR- reads_file - path to the reads file
           to filter. Expects an interleaved file, if it's paired end. If
           both of the above are given, the read_library_ref takes
           precedence.) -> structure: parameter "read_library_ref" of String,
           parameter "reads_file" of String
        :param run_params: instance of type "RQCFilterParams" (Contains all
           parameters for the RQCFilter program, EXCEPT for the inputs and
           outputs. Those are added specifically by each function. This lets
           us describe them separately for the local function that works
           mainly against the file system and the app that mainly works
           against the Workspace. This doesn't cover all of the 110+
           parameters provided by rqcfilter. Those not listed here are left
           as default values, except sketch=f (as that sends data to JGI
           servers for processing), barcodefilter=f, and mapk=13. Notes below
           are taken from the help output from rqcfilter2.sh ver 38.00
           Parameters (format = param name - default - description):
           --------------------------------------------------------- library
           - frag - should be one of 'frag', 'clip', 'lfpe', or 'clrs'.
           Adapter trimming parameters: ----------------------------
           trimfragadapter - f - Trim all known Illumina adapter sequences,
           including TruSeq and Nextera. Quality trimming parameters:
           ---------------------------- qtrim - f - Trim read ends to remove
           bases with quality below minq. Performed AFTER looking for kmers.
           Values: rl (trim both ends), f (neither end), r (right end only),
           l (left end only). trimq - 10 - Trim quality threshold.  Must also
           set qtrim for direction, will be ignored if qtrim=f maxns - 0 -
           Reads with more Ns than this will be discarded. minavgquality - 5
           - (maq) Reads with average quality (before trimming) below this
           will be discarded. minlength - 45 - (ml) Reads shorter than this
           after trimming will be discarded.  Pairs will be discarded only if
           both are shorter. mlf - 0.333 - (minlengthfraction) Reads shorter
           than this fraction of original length after trimming will be
           discarded. Mapping parameters (for vertebrate contaminants):
           ------------------------------------------------- removemouse - f
           - (mouse) Remove mouse reads via mapping. removecat - f - (cat)
           Remove cat reads via mapping. removedog - f - (dog) Remove dog
           reads via mapping. removehuman - f - (human) Remove human reads
           via mapping. Microbial contaminant removal parameters:
           ----------------------------------------- removemicrobes - f -
           (microbes) Remove common contaminant microbial reads via mapping,
           and place them in a separate file. taxlist - emptylist - (tax)
           Remove these taxa from the database before filtering.  Typically,
           this would be the organism name or NCBI ID, or a comma-delimited
           list.  Organism names should have underscores instead of spaces,
           such as Escherichia_coli. Filtering parameters (for artificial and
           microbial contaminants):
           -----------------------------------------------------------------
           rna - f - Remove reads containing RNA-specific artifacts. phix - t
           - Remove reads containing phiX kmers. Clumpify parameters:
           -------------------- clumpify - f - Run clumpify. dedupe - f -
           Remove duplicate reads. opticaldupes - f - Remove optical
           duplicates (Clumpify optical flag). Other processing parameters:
           ---------------------------- khist - f - Set to true to generate a
           kmer-frequency histogram of the output data. (included in report
           in the app, as a file in local function) Memory requirements
           (DON'T EXPOSE THESE TO APPS):
           ------------------------------------------------- maxmem - 50 -
           Set maximum memory flag for RQCFilter to try to allocate. Should
           be an integer, in GB.) -> structure: parameter "library" of
           String, parameter "trimfragadapter" of type "boolean" (A boolean -
           0 for false, 1 for true. @range (0, 1)), parameter "qtrim" of
           String, parameter "trimq" of Long, parameter "maxns" of Long,
           parameter "minavgquality" of Long, parameter "minlength" of Long,
           parameter "mlf" of Double, parameter "removemouse" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "removecat" of type "boolean" (A boolean - 0 for false,
           1 for true. @range (0, 1)), parameter "removedog" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "removehuman" of type "boolean" (A boolean - 0 for
           false, 1 for true. @range (0, 1)), parameter "removemicrobes" of
           type "boolean" (A boolean - 0 for false, 1 for true. @range (0,
           1)), parameter "taxlist" of list of String, parameter "rna" of
           type "boolean" (A boolean - 0 for false, 1 for true. @range (0,
           1)), parameter "phix" of type "boolean" (A boolean - 0 for false,
           1 for true. @range (0, 1)), parameter "clumpify" of type "boolean"
           (A boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "dedupe" of type "boolean" (A boolean - 0 for false, 1 for true.
           @range (0, 1)), parameter "opticaldupes" of type "boolean" (A
           boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "khist" of type "boolean" (A boolean - 0 for false, 1 for true.
           @range (0, 1)), parameter "maxmem" of Long
        :returns: instance of type "RQCFilterLocalOutput" (The output from
           the local function version of RQCFilter. output_directory: the
           path to the output directory containing all files generated by
           RQCFilter. run_log: the path to the run log from RQCFilter (i.e.
           its stderr). This will be a path in the output directory, added
           separately here for convenience. filtered_fastq_file: the path to
           the file (in the output directory) containing the filtered FASTQ
           reads. This will likely be compressed, if you need it
           decompressed, you can use DataFileUtil.unpack_file (see that
           module). run_command: the string that's run on the command line
           with all parameters formatted, etc.) -> structure: parameter
           "output_directory" of String, parameter "run_log" of String,
           parameter "filtered_fastq_file" of String, parameter "run_command"
           of String
        """
        return self._client.call_method('BBTools.run_RQCFilter_local',
                                        [io_params, run_params], self._service_ver, context)

    def run_mem_estimator(self, params, context=None):
        """
        This is a local function that estimates how much memory SPAdes or metaSPAdes needs
        to assemble a paired end library.
        Returns a float, representing the estimated memory use in GB.
        :param params: instance of type "MemEstimatorParams" (reads_file -
           path to a reads file. If this is here alone, expect it to be
           interleaved. reads_file2 - path to the pair of the first file.) ->
           structure: parameter "reads_file" of String, parameter
           "reads_file2" of String
        :returns: instance of type "MemEstimatorOutput" (estimate - the
           estimated amount of memory required to assemble the paired end
           files, in GB. size - the total disk space in GB used by the reads
           files.) -> structure: parameter "estimate" of Double, parameter
           "size" of Double
        """
        return self._client.call_method('BBTools.run_mem_estimator',
                                        [params], self._service_ver, context)

    def bbtools_version(self, context=None):
        """
        Returns the semantic version of the currently installed BBTools. So something like "38.08"
        :returns: instance of String
        """
        return self._client.call_method('BBTools.bbtools_version',
                                        [], self._service_ver, context)

    def status(self, context=None):
        return self._client.call_method('BBTools.status',
                                        [], self._service_ver, context)
